syntax = "proto3";

package kmsp11;

message LibraryConfig {
  // Required. The list of tokens to expose in this library.
  repeated TokenConfig tokens = 1;

  // Optional. The RPC timeout for Cloud KMS API invocations. 0 or unset means
  // the default (30 seconds).
  uint32 rpc_timeout_secs = 2;

  // Optional. The address to invoke. Defaults to `cloudkms.googleapis.com:443`.
  string kms_endpoint = 3;

  // Optional. True means that Google default authentication should be skipped,
  // and insecure gRPC credentials (read: no credentials) should be used
  // instead. Intended for use in test environments only, as this will result in
  // PERMISSION_DENIED in real KMS environments.
  bool use_insecure_grpc_channel_credentials = 4;

  // Optional. True means that X.509 certificates will be generated at runtime
  // for asymmetric KMS keys that are exposed in the library. Certificates are
  // not stable between runs of this library. This functionality is intended for
  // compatibility with the SunPKCS11 JCA provider; other use is discouraged.
  bool generate_certs = 5;

  // Optional. The directory in which log files should be emitted. An empty value
  // means log to standard error.
  string log_directory = 6;
}

message TokenConfig {
  // Required. The Cloud KMS KeyRing associated with this token.
  // For example, projects/foo/locations/global/keyRings/bar
  string key_ring = 1;

  // Optional. The label to assign to this token in the PKCS #11
  // CK_TOKEN_INFO structure.
  string label = 2;
}
